plugin/pkg/admission/webhook/admission.go

```diff
@@ -0,0 +1,220 @@
+	*admission.Handler
+	hookSource           adinit.WebhookSource
+	serviceResolver      adinit.ServiceResolver
+	negotiatedSerializer runtime.NegotiatedSerializer
+	clientCert           []byte
+	clientKey            []byte
+}
+
+var (
+	_ = adinit.WantsServiceResolver(&GenericAdmissionWebhook{})
+	_ = adinit.WantsClientCert(&GenericAdmissionWebhook{})
+	_ = adinit.WantsWebhookSource(&GenericAdmissionWebhook{})
+)
+
+func (a *GenericAdmissionWebhook) SetServiceResolver(sr adinit.ServiceResolver) {
+	a.serviceResolver = sr
+}
+
+func (a *GenericAdmissionWebhook) SetClientCert(cert, key []byte) {
+	a.clientCert = cert
+	a.clientKey = key
+}
+
+func (a *GenericAdmissionWebhook) SetWebhookSource(ws adinit.WebhookSource) {
+	a.hookSource = ws
+}
+
+// Admit makes an admission decision based on the request attributes.
+func (a *GenericAdmissionWebhook) Admit(attr admission.Attributes) (err error) {
+	hooks, err := a.hookSource.List()
+	if err != nil {
+		return fmt.Errorf("failed listing hooks: %v", err)
+	}
+
+	errCh := make(chan error, len(hooks))
+	wg := sync.WaitGroup{}
+	wg.Add(len(hooks))
+	for i := range hooks {
+		go func(hook *admissionregistration.ExternalAdmissionHook) {
+			defer wg.Done()
+			allowed, reason, err := a.callHook(hook, attr)
+			if err != nil {
+				glog.Warningf("Failed calling webhook %v: %v", hook.Name, err)
+				// Since we are failing open to begin with, we do not send an error down the channel
+			} else if !allowed {
+				errCh <- ErrWebhookRejected{
+					WebhookName: hook.Name,
+					Reason:      reason,
+				}
+			}
+		}(&hooks[i])
+	}
+	wg.Wait()
+	close(errCh)
+
+	errs := []error{}
+	for e := range errCh {
+		errs = append(errs, e)
+	}
+	// NewAggregate correctly handles empty lists
+	return utilerrors.NewAggregate(errs)
+}
+
+func (a *GenericAdmissionWebhook) callHook(h *admissionregistration.ExternalAdmissionHook, attr admission.Attributes) (allow bool, reason string, err error) {
+	matches := false
+	for _, r := range h.Rules {
+		m := RuleMatcher{Rule: r, Attr: attr}
+		if m.Matches() {
+			matches = true
+			break
+		}
+	}
+	if !matches {
+		return true, "", nil
+	}
+
+	// Make the webhook request
+	request := admissionv1alpha1.NewAdmissionReview(attr)
+	client, err := a.hookClient(h)
+	if err != nil {
+		return false, "", fmt.Errorf("unable to construct hook %q client: %v", h.Name, err)
+	}
+	response := client.Post().Body(&request).Do()
+
+	// Handle webhook response
+	if err := response.Error(); err != nil {
+		return false, "", err
+	}
+
+	var statusCode int
+	if response.StatusCode(&statusCode); statusCode < 200 || statusCode >= 300 {
+		return false, "", fmt.Errorf("error contacting webhook %q: %d", h.Name, statusCode)
+	}
+
+	if err := response.Into(&request); err != nil {
+		return false, "", fmt.Errorf("error interpreting webhook %q response: %v", h.Name, err)
+	}
+
+	if !request.Status.Allowed {
+		if request.Status.Result != nil {
+			reason = string(request.Status.Result.Reason)
```

Comment:
We should be returning the actual status back, and then converting it to an error.  We shouldn't be transforming status -> string -> error.

I.e. the status field should be passed (if specified) otherwise we need to create a stub one which is generic admission failure.  That status should then be wrapped in StatusErr{}.  I think you could just get away with returning the error here - reason doesn't need to be separately split.